File: client/src/lib/sfu-ice-servers.ts
/**
 * Which ICE servers an SFU (LiveKit) connection is handed, if any.
 *
 * The LiveKit server sends its own ICE list in the join response, and every
 * client used to take that list and nothing else. On the hosted deployment
 * that list is the media box's built-in TURN (UDP on the box itself, and a TLS
 * relay on a port Caddy owns, so dead), while `/api/ice-servers` carries the
 * relays the product actually pays for (Cloudflare first, then Metered, then
 * the static fallback). The mesh path has used that list all along; the SFU
 * path never saw it.
 *
 * Handing the app's list to the SDK is a REPLACEMENT on the web: livekit-client
 * 2.21.0 applies the server's servers only when `rtcConfig.iceServers` is
 * absent (`RTCEngine.makeRTCConfiguration`). So a list without a relay would
 * be strictly worse than today, which is why this returns `undefined`, meaning
 * "pass nothing, keep the server's list", unless the list carries at least one
 * TURN entry. The default STUN-only list in `peer-connection-manager.ts`, an
 * empty list, and a failed fetch all fall in that bucket.
 */
export function sfuIceServers(
  servers: readonly RTCIceServer[] | null | undefined,
): RTCIceServer[] | undefined {
  if (!servers || servers.length === 0) {
    return undefined;
  }
  const hasRelay = servers.some((server) =>
    urlsOf(server).some(isTurnUrl),
  );
  return hasRelay ? [...servers] : undefined;
}

function urlsOf(server: RTCIceServer): string[] {
  const { urls } = server;
  if (typeof urls === "string") {
    return [urls];
  }
  return Array.isArray(urls) ? urls : [];
}

function isTurnUrl(url: string): boolean {
  const scheme = url.trim().toLowerCase();
  return scheme.startsWith("turn:") || scheme.startsWith("turns:");
}

File: ios/pqp/Sources/Voice/SfuIceServers.swift
import Foundation

/// Which ICE servers an SFU (LiveKit) connection is handed, if any.
///
/// The LiveKit server sends its own ICE list in the join response, and this
/// app used to connect with `ConnectOptions()` and take that list and nothing
/// else. On the hosted deployment that list is the media box's built-in TURN
/// (UDP on the box itself, plus a TLS relay on a port Caddy owns, so dead),
/// while `/api/ice-servers` carries the relays the product pays for. The mesh
/// path has used that list all along; the SFU path never saw it.
///
/// Handing the app's list to the SDK is a REPLACEMENT: `client-sdk-swift`
/// 2.16.0 overwrites the server's servers whenever `connectOptions.iceServers`
/// is non-empty (`Room+Engine.swift`, "Override with user provided
/// iceServers"). So a list without a relay would be strictly worse than today,
/// which is why this returns an empty list, meaning "pass nothing, keep the
/// server's list", unless the app's list carries at least one TURN entry. An
/// empty response and a STUN-only response both fall in that bucket. The web
/// client has the same rule in `sfu-ice-servers.ts`.
enum SfuIceServers {
    /// The servers to put in `ConnectOptions(iceServers:)`; empty means none.
    static func select(_ servers: [IceServerConfig]) -> [IceServerConfig] {
        let hasRelay = servers.contains { server in
            server.urlList.contains(where: isTurnUrl)
        }
        return hasRelay ? servers : []
    }

    static func isTurnUrl(_ url: String) -> Bool {
        let scheme = url.trimmingCharacters(in: .whitespaces).lowercased()
        return scheme.hasPrefix("turn:") || scheme.hasPrefix("turns:")
    }
}
